tripy/notebooks/resnet50.ipynb

```diff
@@ -0,0 +1,493 @@
+    "        x = self.convolution(x)\n",
+    "        x = self.normalization(x)\n",
+    "        if self.activation:\n",
+    "            x = tp.relu(x)\n",
+    "        return x\n",
+    "\n",
+    "\n",
+    "class ResNetShortCut(tp.Module):\n",
+    "    def __init__(\n",
+    "        self, in_channels: int, out_channels: int, stride: Tuple[int, int]\n",
+    "    ):\n",
+    "        \"\"\"1x1 convolution and batch normalization for input and output channel\n",
+    "        dimension matching in ResNetBottleNeckLayer.\"\"\"\n",
+    "        super(ResNetShortCut, self).__init__()\n",
+    "        self.convolution = tp.Conv(\n",
+    "            in_channels,\n",
+    "            out_channels,\n",
+    "            kernel_dims=(1, 1),\n",
+    "            stride=stride,\n",
+    "            bias=False,\n",
+    "        )\n",
+    "        self.normalization = tp.BatchNorm(out_channels)\n",
+    "\n",
+    "    def __call__(self, x: tp.Tensor) -> tp.Tensor:\n",
+    "        x = self.convolution(x)\n",
+    "        x = self.normalization(x)\n",
+    "        return x"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class ResNetEmbeddings(tp.Module):\n",
+    "    def __init__(self):\n",
+    "        \"\"\"\n",
+    "        Applies an initial convolution and max-pooling operation to reduce\n",
+    "        dimensions and prepare input for the encoder.\n",
+    "        \"\"\"\n",
+    "        super(ResNetEmbeddings, self).__init__()\n",
+    "        self.embedder = ResNetConvLayer(\n",
+    "            3,\n",
+    "            64,\n",
+    "            kernel_dims=(7, 7),\n",
+    "            stride=(2, 2),\n",
+    "            padding=((3, 3), (3, 3)),\n",
+    "        )\n",
+    "\n",
+    "    def __call__(self, x: tp.Tensor) -> tp.Tensor:\n",
+    "        x = self.embedder(x)\n",
+    "        x = tp.maxpool(\n",
+    "            x,\n",
+    "            kernel_dims=(3, 3),\n",
+    "            stride=(2, 2),\n",
+    "            padding=((1, 1), (1, 1)),\n",
+    "        )\n",
+    "        return x\n",
+    "\n",
+    "\n",
+    "class ResNetBottleNeckLayer(tp.Module):\n",
+    "    def __init__(\n",
+    "        self,\n",
+    "        in_channels: int,\n",
+    "        out_channels: int,\n",
+    "        bottleneck_channels: int,\n",
+    "        stride: Tuple[int, int],\n",
+    "    ):\n",
+    "        \"\"\"Consists of a ResNetShortCut layer for dimension matching and\n",
+    "        three ResNetConvLayers for residual learning.\"\"\"\n",
+    "        super(ResNetBottleNeckLayer, self).__init__()\n",
+    "\n",
+    "        self.shortcut = (\n",
+    "            ResNetShortCut(in_channels, out_channels, stride)\n",
+    "            if in_channels != out_channels or stride != (1, 1)\n",
+    "            else lambda x: x\n",
+    "        )\n",
+    "        self.layer = [\n",
```

Comment:
Should we replace all the lists with `tp.Sequential`?

Reply:
I think not, because sequential is one big forward call. But usually you want to do

```python
for layer in layers:
    x= layer(x)
```
 
 instead of `x=layers(x)`. It is also good to have both to demonstrate functionality options.

Reply:
switched to sequential, since it is the same input for all layers.